doc/whats_new/v1.2.rst

```diff
@@ -138,6 +138,9 @@ Changelog
   :class:`cluster.AgglomerativeClustering` and will be renamed to `metric` in v1.4.
   :pr:`23470` by :user:`Meekail Zain <micky774>`.
 
+- |Enhancement| The `_k_means_lloyd` routine no longer needs the unused `x_squared_norms`
+  argument. :pr:`24264` by :user:`Vincent Maladiere <Vincent-Maladiere>`.
```

Comment:
We typically include changes to public interfaces in the changelog. In this PR, removing `x_squared_norms` is a internal maintenance fix.

Although, I do see that `KMeans.predict` does not call `row_norms` anymore, so this PR may become an `|Efficiency|` change. May you do a quick benchmark to check that `KMean.predict` is faster in this PR compared to `main`?

Reply:
Hello Thomas! Thank you for your feedback. I will remove these lines from the changelog and run a quick benchmark :)

Reply:
The benchmark didn't yield statistically significant results between `main` and this PR.

### main

- Mean

| init      | time     | inertia      | homo     | compl    | v-meas   | ARI      | AMI      | silhouette |
|-----------|----------|--------------|----------|----------|----------|----------|----------|------------|
| k-means++ | 0.070214 | 69662.003524 | 0.679688 | 0.718645 | 0.698624 | 0.570117 | 0.695492 | 0.177128   |
| pca       | 0.017836 | 74152.347049 | 0.653272 | 0.682705 | 0.667664 | 0.538763 | 0.664255 | 0.138850   |
| random    | 0.055914 | 69707.411869 | 0.674647 | 0.715533 | 0.694489 | 0.559928 | 0.691309 | 0.175797   |

- Standard deviation

| init      | time     | inertia      | homo         | compl        | v-meas       | ARI | AMI          | silhouette |
|-----------|----------|--------------|--------------|--------------|--------------|-----|--------------|------------|
| k-means++ | 0.019954 | 6.176944e-12 | 0.000000e+00 | 0.000000e+00 | 0.000000e+00 | 0.0 | 0.000000e+00 | 0.009096   |
| pca       | 0.002190 | 3.222816e-12 | 9.293440e-18 | 1.405036e-17 | 1.756295e-17 | 0.0 | 2.249155e-17 | 0.009011   |
| random    | 0.009075 | 4.604022e-13 | 0.000000e+00 | 0.000000e+00 | 0.000000e+00 | 0.0 | 0.000000e+00 | 0.009588   |


### This PR

- Mean

| init      | time     | inertia      | homo     | compl    | v-meas   | ARI      | AMI      | silhouette |
|-----------|----------|--------------|----------|----------|----------|----------|----------|------------|
| k-means++ | 0.075397 | 69662.003524 | 0.679688 | 0.718645 | 0.698624 | 0.570117 | 0.695492 | 0.177218   |
| pca       | 0.016281 | 72686.225300 | 0.636237 | 0.658148 | 0.647007 | 0.520614 | 0.643407 | 0.143424   |
| random    | 0.056225 | 69707.411869 | 0.674647 | 0.715533 | 0.694489 | 0.559928 | 0.691309 | 0.175988   |

- Standard deviation

| init      | time     | inertia      | homo | compl | v-meas | ARI | AMI | silhouette |
|-----------|----------|--------------|------|-------|--------|-----|-----|------------|
| k-means++ | 0.015151 | 5.913974e-12 | 0.0  | 0.0   | 0.0    | 0.0 | 0.0 | 0.009003   |
| pca       | 0.005357 | 0.000000e+00 | 0.0  | 0.0   | 0.0    | 0.0 | 0.0 | 0.008478   |
| random    | 0.008589 | 1.029491e-12 | 0.0  | 0.0   | 0.0    | 0.0 | 0.0 | 0.009005   |


<br>

I followed [this scikit-learn benchmark](https://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_digits.html).
See the full code on [this gist](https://gist.github.com/Vincent-Maladiere/e08b533fee5e2b2a214e5fb23bb67019).

Reply:
I suspect the digits dataset is not big enough to see a difference in `row_norms`. Here is a quick benchmark using a bigger dataset:

```python
from sklearn.datasets import make_blobs
from sklearn.cluster import KMeans

X, _ = make_blobs(n_samples=100_000, n_features=100, random_state=0)
kmeans = KMeans(n_init="auto").fit(X)
```

### This PR

```python
%%timeit
_ = kmeans.predict(X)
# 8.86 ms ± 491 µs per loop
```

### On main

```python
%%timeit
_ = means.predict(X)
# 10.8 ms ± 372 µs per loop
```

Here you see that this PR is faster. This is expected because there is a ~2ms overhead from running `row_norms`.

```python
from sklearn.utils.extmath import row_norms

X, _ = make_blobs(n_samples=100_000, n_features=100, random_state=0)

%%timeit
_ = row_norms(X, squared=True)
# 2.21 ms ± 9.79 µs per loop
```

Reply:
Can you see what happens with your benchmark when the data is `make_blobs(n_samples=100_000, n_features=100, random_state=0)`?

Reply:
I still got non-significant results, something must be wrong with my benchmark.

I observe a perf gain with your experiment, however.

### This PR
```python
%%timeit
_ = means.predict(X)
# 10.9 ms ± 133 µs per loop
```

### On main
```python
%%timeit
_ = means.predict(X)
# 18.9 ms ± 276 µs per loop
```

Reply:
May you add an changelog entry for the speed improvement with the `|Efficiency|` tag?